### PR TITLE
Fixed incorrect suggestion for object initialization simplification

### LINKED ISSUE #23368: Incorrect suggestion for object initialization simplification
Consider the following code:

``` c#
interface IExample {
    string Name { get; set; }
}

class C : IExample {
    string IExample.Name { get; set; }
}

void Repro() {
  IExample e = new C();
  e.Name = "jared"
}
```

Given this code the IDE will add the suggestion IDE0017, object initialization can be simplified, on the `IExample e` line. The suggested fix is the following:

``` c#
IExample e = new C {
  Name = "jared"
};
```

This results in code that won't compile. The name `Name` is only legal through the type `IExample` but the fix causes `Name` to be evaluated in the context of `C` instead.

### PR DESCRIPTION
**Pull request by zaytsev-victor**
**Customer scenario**

User can see suggestion for object initialization simplification even if result won't compile.

**Bugs this fixes:**
Fixes #23368

**Risk**
Low

**Performance impact**
Low

**Is this a regression from a previous update?**
No

### CODE REVIEW

File: src/Features/Core/Portable/UseObjectInitializer/ObjectCreationExpressionAnalyzer.cs [hunk 1]
```diff
@@ -164,6 +170,18 @@ protected override void AddMatches(ArrayBuilder<Match<TExpressionSyntax, TStatem
             }
         }
 
+        private static bool IsExplicitlyImplemented(
+            INamedTypeSymbol classOrStructType,
+            ISymbol member)
+        {
+            var implementation = classOrStructType?.FindImplementationForInterfaceMember(member);
+            switch (implementation)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UseObjectInitializer/ObjectCreationExpressionAnalyzer.cs]: swithc seems like overkill here.  why not just:

```c#
return implementation is IPropertySymbol property && ...
```

File: src/Features/Core/Portable/UseObjectInitializer/ObjectCreationExpressionAnalyzer.cs [hunk 2]
```diff
@@ -175,11 +175,9 @@ protected override void AddMatches(ArrayBuilder<Match<TExpressionSyntax, TStatem
...
         {
             var implementation = classOrStructType?.FindImplementationForInterfaceMember(member);
-            switch (implementation)
-            {
-                case IPropertySymbol property: return property.ExplicitInterfaceImplementations.Length > 0 && property.DeclaredAccessibility == Accessibility.Private;
-                default: return false;
-            }
+            return implementation is IPropertySymbol property && 
+                property.ExplicitInterfaceImplementations.Length > 0 
+                && property.DeclaredAccessibility == Accessibility.Private;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UseObjectInitializer/ObjectCreationExpressionAnalyzer.cs]: && on previous line.

**Comment by Neme12** [comment]
Why do all these PRs keep getting delayed :(

**Comment by zaytsev-victor** [comment]
@jcouv @sharwell any update on this?

**Comment by jcouv** [comment]
@dotnet/roslyn-ide Can we get a final review on this community PR? Thanks

File: src/EditorFeatures/CSharpTest/UseObjectInitializer/UseObjectInitializerTests.cs [hunk 3]
```diff
@@ -576,6 +576,106 @@ public void Main()
...
+    string IExample.Name { get; set; }
+    public string LastName { get; set; }
+}
+
+class MyClass
+{
+    public void Main()
+    {
+        IExample e = [||]new C();
+        e.Name = string.Empty;
```
Comment (heejaechang) [inline src/EditorFeatures/CSharpTest/UseObjectInitializer/UseObjectInitializerTests.cs]: so the order causing one to work and not the other? would it be possible to support both if possible?

File: src/EditorFeatures/VisualBasicTest/UseObjectInitializer/UseObjectInitializerTests.vb [hunk 4]
```diff
@@ -369,6 +369,108 @@ Class C
...
+
+        <WorkItem(23368, "https://github.com/dotnet/roslyn/issues/23368")>
+        <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseObjectInitializer)>
+        Public Async Function TestWithExplicitImplementedInterfaceMembers3() As Task
+            Await TestInRegularAndScriptAsync(
+"
+class C
+    Sub Bar()
+        Dim c As IExample = [||]New Foo
+        c.LastName = String.Empty
```
Comment (heejaechang) [inline src/EditorFeatures/VisualBasicTest/UseObjectInitializer/UseObjectInitializerTests.vb]: same here.

**Comment by heejaechang** [comment]
retest this please

File: src/Features/Core/Portable/UseObjectInitializer/ObjectCreationExpressionAnalyzer.cs [hunk 5]
```diff
@@ -164,6 +170,16 @@ protected override void AddMatches(ArrayBuilder<Match<TExpressionSyntax, TStatem
...
         }
 
+        private static bool IsExplicitlyImplemented(
+            INamedTypeSymbol classOrStructType,
+            ISymbol member)
+        {
+            var implementation = classOrStructType?.FindImplementationForInterfaceMember(member);
+            return implementation is IPropertySymbol property && 
+                property.ExplicitInterfaceImplementations.Length > 0 && 
+                property.DeclaredAccessibility == Accessibility.Private;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UseObjectInitializer/ObjectCreationExpressionAnalyzer.cs]: note: vb can have public implementations where the member name does not match.

File: src/EditorFeatures/VisualBasicTest/UseObjectInitializer/UseObjectInitializerTests.vb [hunk 6]
```diff
@@ -369,6 +369,108 @@ Class C
...
+Interface IExample
+    Property Name As String
+    Property LastName As String
+End Interface
+
+Class Foo
+    Implements IExample
+
+    Private Property Name As String Implements IExample.Name
+    Public Property LastName As String Implements IExample.LastName
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/VisualBasicTest/UseObjectInitializer/UseObjectInitializerTests.vb]: can we have a test where these names don't match.  i.e. "Public Property MyLastName Implements IExample.LastName"?

**Comment by jinujoseph** [comment]
@zaytsev-victor could you resolve the conflict and address cyrus feedback pls.

File: src/EditorFeatures/VisualBasicTest/UseObjectInitializer/UseObjectInitializerTests.vb [hunk 7]
```diff
@@ -471,6 +471,55 @@ Class Foo
...
+
+    Private Property Name As String Implements IExample.Name
+    Public Property MyLastName As String Implements IExample.LastName
+End Class
+",
+"
+class C
+    Sub Bar()
+        Dim c As IExample = New Foo With {
+            .LastName = String.Empty
```
Comment (heejaechang) [inline src/EditorFeatures/VisualBasicTest/UseObjectInitializer/UseObjectInitializerTests.vb]: this is error? can we make it to use MyLastName ?

**Comment by jinujoseph** [comment]
@CyrusNajmabadi  can you review if your feedback is addressed

**Comment by CyrusNajmabadi** [comment]
Yup.  I will take a look later.  For some reason my github rules are not working and i'm not getting notifications about these directly :-/

File: src/Features/VisualBasic/Portable/UseObjectInitializer/VisualBasicUseObjectInitializerCodeFixProvider.vb [hunk 8]
```diff
@@ -67,7 +67,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.UseObjectInitializer
                 Dim initializer = SyntaxFactory.NamedFieldInitializer(
                     keyKeyword:=Nothing,
                     dotToken:=match.MemberAccessExpression.OperatorToken,
-                    name:=DirectCast(match.MemberAccessExpression.Name, IdentifierNameSyntax),
+                    name:=SyntaxFactory.IdentifierName(match.MemberName),
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/UseObjectInitializer/VisualBasicUseObjectInitializerCodeFixProvider.vb]: this likely will not work for names that are keywords.  You store .ValueText which won't have hte propery escaping here.

Comment (zaytsev-victor) [reply src/Features/VisualBasic/Portable/UseObjectInitializer/VisualBasicUseObjectInitializerCodeFixProvider.vb]: @CyrusNajmabadi I've tested it with ```Class```, ```Property``` and ```Date``` keywords. It works, probably we don't need to escape names in object initializer.

File: src/Features/Core/Portable/UseObjectInitializer/ObjectCreationExpressionAnalyzer.cs [hunk 9]
```diff
@@ -160,10 +166,28 @@ protected override void AddMatches(ArrayBuilder<Match<TExpressionSyntax, TStatem
...
+            out ISymbol typeMember)
+        {
+            if (member != null && member.ContainingType.IsInterfaceType())
+            {
+                typeMember = classOrStructType?.FindImplementationForInterfaceMember(member);
+                return typeMember is IPropertySymbol property &&
+                    property.ExplicitInterfaceImplementations.Length > 0 &&
+                    property.DeclaredAccessibility == Accessibility.Private;
+                
+            }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UseObjectInitializer/ObjectCreationExpressionAnalyzer.cs]: extra newline above this.

**Comment by zaytsev-victor** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by jinujoseph** [comment]
approved to merge for 15.8.Preview4

**Pull request merged**